README
xpbres - extended pbres

last versions you can find at http://fedorchenko.net/xpbres.php


Changelog
22.05.2010
    - xpbres-0.2
    - add header file write for converted ibitmap
30.04.2010
    - cpbres-0.1
    - xpm format support

pbres.h
/*
 * Yury P. Fedorchenko 2010
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor Boston, MA 02110-1301,  USA
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <zlib.h>

#define MAXSIZE 520000
#define MAXIMGLIST 4096
#define PBTSIGNATURE "PocketBookTheme"
#define PBTVERSION 1

typedef struct ibitmap_s {

	unsigned short width;
	unsigned short height;
	unsigned short depth;
	unsigned short scanline;
	unsigned char data[];

} ibitmap;

struct imgcache {

	int pos;
	int len;
	unsigned char *data;

};
#ifdef __cplusplus
extern "C" {
#endif
void warning(const char *fmt, ...);
void terminate(const char *fmt, ...);
void usage(void);
void write_as_c(FILE *f, unsigned char *data, int len);
ibitmap *bmp2res(const char *fname, unsigned char *data, int len, int *outsize);
ibitmap *xpm2res(const char *fname, unsigned char *data, int len, int *outsize);
char *getname(const char *s);
#ifdef __cplusplus
};
#endif


pbxpm.cc
/*
 * Yury P. Fedorchenko 2010
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor Boston, MA 02110-1301,  USA
 */

#include "pbres.h"
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <sstream>
#include <iostream>

struct mxpm{
  int w,h,nc;
  bool read,pal_read,finish;
  std::map<char,std::string> pal;
  std::map<char,unsigned char> pal_pb;
  std::vector<std::string> lines;
  mxpm():read(false),pal_read(false),finish(false){}
};


unsigned char pal2pb(const std::string& ss){
    std::string s;
    s+='0';s+=ss;
    int val;
    sscanf(s.c_str(),"%x",&val);
    unsigned char* pos=reinterpret_cast<unsigned char* >(&val);
    
    unsigned char c = (pos[0] + pos[1]*6 + pos[2]*3) / 10;
    if (c <= 0x2a) {
	return 0;
    } else if (c <= 0x80) {
	return 1;
   } else if (c <= 0xd4) {
	return 2;
   } else {
	return 3;
   }
}

std::ostream& operator<<(std::ostream& s,const mxpm& p){
  for(std::map<char,std::string>::const_iterator it=p.pal.begin();it!=p.pal.end();++it){
    s<<it->first<<" c "<<it->second<<std::endl;
  }
  for(std::map<char,unsigned char>::const_iterator it=p.pal_pb.begin();it!=p.pal_pb.end();++it){
    s<<it->first<<" c "<<(int)it->second<<std::endl;
  }
  for(std::vector<std::string>::const_iterator it=p.lines.begin();it!=p.lines.end();++it){
    s<<*it<<std::endl;
  }
  return s;
}

ibitmap *xpm2res(const char *fname, unsigned char* data,int len,int *outsize) {
	std::cerr<<"Processing:"<<fname<<std::endl;
	ibitmap *bm;
	unsigned char *udata, *umask, *outp, *maskp;
	int i, x, y, usemask;

#ifndef CONVERT
	std::stringstream in((const char*)data);
#else
	std::ifstream in(fname);
#endif
	mxpm px;
	while(!in.eof() && !in.fail()){
	  std::string str;
	  std::getline(in,str);
	  size_t ff=str.find_first_of("\"");
	  if(ff!=str.npos){
	    if(!px.read){
	      std::stringstream ss(str.substr(ff+1));
	      ss>>px.w>>px.h>>px.nc;
	      if(px.nc==0)terminate("error in xpm file %s %s\n",fname,str.c_str());
	      else px.read=true;
	    }else if(!px.pal_read){
	      std::string ss(str.substr(ff+1));
	      char cn,empt;
	      std::string rest;
	      cn=ss[0];
//	      ss>>cn>>empt>>rest;
	      rest=ss.substr(ss.find_last_of("#N"));
	      px.pal[cn]=rest.substr(0,rest.find_last_of("\""));
	      if(px.pal[cn]=="None" || px.pal[cn]=="Null")px.pal_pb[cn]=255;
	      else{rest=px.pal[cn];rest[0]='x';
	      px.pal_pb[cn]=pal2pb(rest);}
	      if(px.pal.size()==px.nc)px.pal_read=true;
	    }else{
	      px.lines.push_back(str.substr(ff+1,str.find_last_of("\"")));
	      if(px.lines.size()==px.h){
	      px.finish=true;
	      break;
	      }
	    }
	  }
	}
	
	int scanout = (px.w + 3) / 4;
	int maskout = (px.w + 7) / 8;


	int ulen = px.h * scanout;
	int umasklen = px.h * maskout;

	bm = (ibitmap*)calloc(sizeof(ibitmap) + ulen + umasklen +10, 1);
	bm->width = px.w;
	bm->height = px.h;
	bm->depth = 2;//(bpp == 1) ? 1 : 2;
	bm->scanline = scanout;

	udata = bm->data;
	memset(udata, 0, ulen);
	umask = bm->data + ulen;
	memset(umask, 0xff, umasklen);
	usemask = 0;

	for (y=0; y<px.h; y++) {

		outp = udata + (y) * scanout;
		maskp = umask + (y) * maskout;

		/*if (bpp == 1) {

			inv = (palette[0] < palette[1]) ? 0xff : 0;
			for (i=0; i<scanout; i++) {
				outp[i] = inp[i] ^ inv;
			}

		} else {*/

			for (x=0; x<px.w; x++) {
				if (px.pal_pb[px.lines[y][x]] != 255) {
					outp[x>>2] |= (px.pal_pb[px.lines[y][x]]) << ((3 - (x & 3)) << 1);
				} else {
					usemask = 1;
					maskp[x>>3] &= ~(1 << (7 - (x & 7)));
				}

			}

		//}

	}

	if (usemask) bm->depth |= 0x8000;
	*outsize = usemask ? ulen+umasklen : ulen;
	std::cerr<<"end Processing:"<<fname<<std::endl;
	return bm;

}

#ifdef CONVERT
void write_as_c(FILE *f, unsigned char *data, int len) {

	char buf[512];
	int i, n, ll;

	while (len > 0) {

		n = (len > 16) ? 16 : len;

		ll = 0;
		ll += sprintf(buf+ll, "\t\t");
		for (i=0; i<n; i++) ll += sprintf(buf+ll, "0x%02x,", data[i]);
		ll += sprintf(buf+ll, "\n");

		if (fwrite(buf, 1, ll, f) != ll) terminate("error writing to file");

		data += n;
		len -= n;

	}

}
void terminate(const char *fmt, ...) {

	va_list ap;

	fprintf(stderr, "Error: ");
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	fprintf(stderr, "\n\n");
	exit(1);

}
char *getname(char *s) {

	static char tname[64];
	char *p;

	p = strrchr(s, '/');
	if (p == NULL) p = strrchr(s, '\\');
	if (p != NULL) p++;
	if (p == NULL) p = s;
	strncpy(tname, p, 63);
	p = strchr(tname, '.');
	if (p != NULL) *p = 0;
	return tname;

}

void usage(const char* nm){
  std::cerr<<"usage"<<std::endl;
  exit(1);
}

int main(int argc,char** argv){
  int beg;
  FILE* out=stdout;
  if(argc<2)usage(argv[0]);
  if(strcmp(argv[1],"-o")==0){
    if(argc>4)
    out=fopen(argv[2],"w");
    else
    usage(argv[0]);
  }
  for(int i=1;i<argc;++i){
    int size;
    ibitmap *bm=xpm2res(argv[i],NULL,0,&size);
    fprintf(out, "const ibitmap %s = {\n", getname(argv[i]));
		fprintf(out, "\t%u, %u, %u, %u,\n", bm->width, bm->height, bm->depth, bm->scanline);
		fprintf(out, "\t{\n");
		write_as_c(out, bm->data, size);
		fprintf(out, "\t}\n");
		fprintf(out, "};\n\n");

  }
  return 0;
}
#endif


pbres.c
/* 
 *  pbres - theme and resource compiler for PocketBook
 *  Author: Dmitry Zakharov
 *  License: GPL
 *  modified by Yury P. Fedorchenko 2010
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <zlib.h>

#define MAXSIZE 520000
#define MAXIMGLIST 4096
#define PBTSIGNATURE "PocketBookTheme"
#define PBTVERSION 1

typedef struct ibitmap_s {

	unsigned short width;
	unsigned short height;
	unsigned short depth;
	unsigned short scanline;
	unsigned char data[];

} ibitmap;

struct imgcache {

	int pos;
	int len;
	unsigned char *data;

};

void warning(const char *fmt, ...) {

	va_list ap;

	fprintf(stderr, "Warning: ");
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	fprintf(stderr, "\n\n");

}

void terminate(const char *fmt, ...) {

	va_list ap;

	fprintf(stderr, "Error: ");
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	fprintf(stderr, "\n\n");
	exit(1);

}

void usage(char **argv) {

	fprintf(stderr, "Usage: %s -c output.c resource [ ... ]\n",argv[0]);
	fprintf(stderr, "       %s -t output.pbt config.txt resource [ ... ]\n",argv[0]);
	fprintf(stderr, "       %s -u input.pbt resource\n",argv[0]);
	fprintf(stderr, "       %s -l input.pbt\n",argv[0]);
	exit(1);

}

void write_as_c(FILE *f, unsigned char *data, int len) {

	char buf[512];
	int i, n, ll;

	while (len > 0) {

		n = (len > 16) ? 16 : len;

		ll = 0;
		ll += sprintf(buf+ll, "\t\t");
		for (i=0; i<n; i++) ll += sprintf(buf+ll, "0x%02x,", data[i]);
		ll += sprintf(buf+ll, "\n");

		if (fwrite(buf, 1, ll, f) != ll) terminate("error writing to file");

		data += n;
		len -= n;

	}

}

ibitmap *bmp2res(const char *fname, unsigned char *data, int len, int *outsize) {

	ibitmap *bm;
	unsigned char palette[256];
	unsigned char *pos, c, inv;
	int start, hsize, width, height, bpp, comp, scanlen, scanout, maskout;
	unsigned char *udata, *umask, *inp, *outp, *maskp;
	int ulen, umasklen;
	int i, x, y, usemask;

	if (data[0] != 'B' || data[1] != 'M') terminate("%s is not a bmp file", fname);

	start = *((int *)(data+0x0a));
	hsize = *((int *)(data+0x0e));
	width = *((int *)(data+0x12));
	height = *((int *)(data+0x16));
	bpp = *((short *)(data+0x1c));
	comp = *((int *)(data+0x1e));
	scanlen = ((width * bpp + 7) / 8 + 3) & ~0x3;
	if (bpp == 1) {
		scanout = (width + 7) / 8;
	} else {
		scanout = (width + 3) / 4;
	}
	maskout = (width + 7) / 8;

	if (comp != 0) terminate("%s: compressed bmp files are not supported", fname);

	if (bpp != 1 && bpp != 8) terminate("%s: wrong format (should be 8-bit or 1-bit)", fname);

	// reading palette

	pos = data + 14 + hsize;
	for (i=0; i<256; i++) {
		if (pos > data+len) break;
		c = (pos[0] + pos[1]*6 + pos[2]*3) / 10;
		if (c <= 0x2a) {
			palette[i] = 0;
		} else if (c <= 0x80) {
			palette[i] = 1;
		} else if (c <= 0xd4) {
			palette[i] = 2;
		} else {
			palette[i] = 3;
		}
		if (pos[0] == 64 && pos[1] == 128 && pos[2] == 128) palette[i] = 255; // transparent
		pos += 4;
	}

	ulen = height * scanout;
	umasklen = height * maskout;

	bm = calloc(sizeof(ibitmap) + ulen + umasklen, 1);
	bm->width = width;
	bm->height = height;
	bm->depth = (bpp == 1) ? 1 : 2;
	bm->scanline = scanout;

	udata = bm->data;
	memset(udata, 0, ulen);
	umask = bm->data + ulen;
	memset(umask, 0xff, umasklen);
	usemask = 0;

	for (y=0; y<height; y++) {

		inp = data + start + y * scanlen;
		outp = udata + (height - y - 1) * scanout;
		maskp = umask + (height - y - 1) * maskout;

		if (bpp == 1) {

			inv = (palette[0] < palette[1]) ? 0xff : 0;
			for (i=0; i<scanout; i++) {
				outp[i] = inp[i] ^ inv;
			}

		} else {

			for (x=0; x<width; x++) {
				if (palette[inp[x]] != 255) {
					outp[x>>2] |= palette[inp[x]] << ((3 - (x & 3)) << 1);
				} else {
					usemask = 1;
					maskp[x>>3] &= ~(1 << (7 - (x & 7)));
				}
			}

		}

	}

	if (usemask) bm->depth |= 0x8000;
	*outsize = usemask ? ulen+umasklen : ulen;
	return bm;

}

char *getname(const char *s) {

	static char tname[64];
	char *p;

	p = strrchr(s, '/');
	if (p == NULL) p = strrchr(s, '\\');
	if (p != NULL) p++;
	if (p == NULL) p = s;
	strncpy(tname, p, 63);
	p = strchr(tname, '.');
	if (p != NULL) *p = 0;
	return tname;

}

void main_c(const char *outfile, int nfiles, const char **infiles) {

	unsigned char *data;
	char *ext,ohname[1024],*outname;
	int i, len, size;
	ibitmap *bm;
	FILE *ifd, *ofd, *ofdh;

	data = malloc(MAXSIZE);
        strcpy(ohname,outfile);
        ohname[strlen(ohname)-1]='\0';
        strcat(ohname,"h");
        ofdh= fopen(ohname, "wb");
	ofd = fopen(outfile, "wb");
	if (ofd == NULL) terminate("Cannot open output file %s", outfile);

	fprintf(ofd, "typedef struct ibitmap_s {\n");
        fprintf(ofd, "\tshort width;\n");
        fprintf(ofd, "\tshort height;\n");
        fprintf(ofd, "\tshort depth;\n");
        fprintf(ofd, "\tshort scanline;\n");
        fprintf(ofd, "\tunsigned char data[];\n");
	fprintf(ofd, "} ibitmap;\n\n");

	for (i=0; i<nfiles; i++) {

		ifd = fopen(infiles[i], "rb");
		if (ifd == NULL) terminate("Cannot open %s", infiles[i]);
		memset(data, 0, MAXSIZE);
		len = fread(data, 1, MAXSIZE, ifd);
		data[len]=0;
		if (len == MAXSIZE) terminate("File %s is too big", infiles[i]);
		fclose(ifd);
		ext=strrchr(infiles[i],'.');
                if(ext && strcasecmp(ext,".bmp")==0 )
		    bm = bmp2res(infiles[i], data, len, &size);
		else bm = xpm2res(infiles[i], data, len, &size);
		outname = getname(infiles[i]);
		fprintf(ofd, "const ibitmap %s = {\n", outname);
		fprintf(ofd, "\t%u, %u, %u, %u,\n", bm->width, bm->height, bm->depth, bm->scanline);
		fprintf(ofd, "\t{\n");
		write_as_c(ofd, bm->data, size);
		fprintf(ofd, "\t}\n");
		fprintf(ofd, "};\n\n");
		//free(bm);
		//fprintf(ofdh, "extern const ibitmap %s;\n", outname);
		fprintf(ofdh, "extern ibitmap %s;\n", outname);

	}

}

void main_t(char *outfile, char *cfgfile, int nfiles, char **infiles) {

	unsigned char *data, *data1, *data2, *header, *hpos;
	unsigned int *iheader;
	int i, j, len, size, headersize, the_same;
	unsigned long clen;
	ibitmap *bm;
	FILE *ifd, *ofd;
	char *fname, *extp;
	struct imgcache *imglist;
	int imgpos;

	data = malloc(MAXSIZE);
	imglist = (struct imgcache *) malloc(MAXIMGLIST * sizeof(struct imgcache));

	ofd = fopen(outfile, "wb");
	if (ofd == NULL) terminate("Cannot open output file %s", outfile);

	ifd = fopen(cfgfile, "rb");
	if (ifd == NULL) terminate("Cannot open configuration file %s", cfgfile);
	len = fread(data, 1, MAXSIZE, ifd);
	if (len == MAXSIZE) terminate("File %s is too big", cfgfile);
	data[len++] = 0;
	clen = len + 16384;
	data2 = malloc(clen);
	compress2 (data2, &clen, data, len, 9);

	headersize = 32;
	for (i=0; i<nfiles; i++) {
		headersize += 12;
		headersize += ((strlen(getname(infiles[i])) / 4) + 1) * 4;
	}

	header = (unsigned char *) malloc(headersize);
	iheader = (int *) header;
	memset(header, 0, headersize);
	strcpy(header, PBTSIGNATURE);
	header[15] = PBTVERSION;
	iheader[4] = headersize;
	iheader[5] = len;
	iheader[6] = headersize;
	iheader[7] = clen;

	fwrite(header, 1, headersize, ofd);
	fwrite(data2, 1, clen, ofd);
	free(data2);

	hpos = header+32;

	imgpos = 0;

	for (i=0; i<nfiles; i++) {

		fname = infiles[i];
		extp = fname + strlen(fname) - 4;
		if (strcasecmp(fname, "thumbs.db") == 0) continue;

		ifd = fopen(fname, "rb");
		if (ifd == NULL) terminate("Cannot open %s", fname);
		memset(data, 0, MAXSIZE);
		len = fread(data, 1, MAXSIZE, ifd);
		if (len == MAXSIZE) terminate("File %s is too big", fname);
		fclose(ifd);

		if (strcasecmp(extp, ".bmp") == 0) {

			bm = bmp2res(fname, data, len, &size);
			data1 = (unsigned char *) bm;
			len = sizeof(ibitmap) + size;

		} else if (strcasecmp(extp, ".ttf") == 0) {

			data1 = data;

		} else {

			data1 = NULL;
			terminate("%s: not a BMP or TTF resource", fname);

		}

		clen = len + 16384;
		data2 = malloc(clen);
		compress2 (data2, &clen, data1, len, 9);


		*((int *) hpos) = len;
		*((int *) (hpos+4)) = ftell(ofd);
		*((int *) (hpos+8)) = clen;

		the_same = 0;
		for (j=0; j<imgpos; j++) {
			if (clen == imglist[j].len && memcmp(data2, imglist[j].data, clen) == 0) {
				*((int *) (hpos+4)) = imglist[j].pos;
				the_same = 1;
				break;
			}
		}
		hpos += 12;
		strcpy(hpos, getname(fname));
		hpos += ((strlen(getname(fname)) / 4) + 1) * 4;

		if (! the_same) {

			if (imgpos < MAXIMGLIST) {
				imglist[imgpos].pos = ftell(ofd);
				imglist[imgpos].len = clen;
				imglist[imgpos].data = malloc(clen);
				memcpy(imglist[imgpos].data, data2, clen);
				imgpos++;
			}
			if (fwrite(data2, 1, clen, ofd) != clen) {
				terminate("Error writing output file");
			}

		}

		free(data2);

	}

	fseek(ofd, 0, SEEK_SET);
	fwrite(header, 1, headersize, ofd);
	fclose(ofd);

}

void unpack_resource(FILE *fd, char *name, unsigned long len, int pos, unsigned long clen) {

	unsigned char *data, *cdata;
	FILE *ofd;
	int r;

	cdata = malloc(clen);
	data = malloc(len+16);
	memset(cdata, 0, clen);
	memset(data, 0, len+16);
	fseek(fd, pos, SEEK_SET);
	fread(cdata, 1, clen, fd);
	r = uncompress (data, &len, cdata, clen);
	if (r != Z_OK) {
		terminate("decompression error");
	}
	ofd = fopen(name, "wb");
	if (ofd == NULL) terminate("Cannot open output file %s", name);
	fwrite(data, 1, len, ofd);
	fclose(ofd);
	free(cdata);
	free(data);

}

void main_u(char *themefile, char *resource) {

	char buf[32];
	unsigned char *header, *hpos;
	unsigned int *iheader;
	int pos, headersize;
	unsigned long len, clen;
	FILE *tfd;

	tfd = fopen(themefile, "rb");
	if (tfd == NULL) terminate("Cannot open theme file %s", themefile);

	memset(buf, 0, 32);
	fread(buf, 1, 32, tfd);
	if (strncmp(buf, PBTSIGNATURE, strlen(PBTSIGNATURE)) != 0) terminate("%s is not a PocketBook theme file");
	headersize = *((int *) (buf+16));
	header = malloc(headersize);
	iheader = (int *) header;
	fseek(tfd, 0, SEEK_SET);
	fread(header, 1, headersize, tfd);

	hpos = header+32;

	if (strcmp(resource, "-") == 0) {
		unpack_resource(tfd, "theme.cfg", iheader[5], iheader[6], iheader[7]);
		return;
	}

	while (hpos < header+headersize) {

		if (strcmp(hpos+12, resource) == 0) {

			len = *((int *) hpos);
			pos = *((int *) (hpos+4));
			clen = *((int *) (hpos+8));
			unpack_resource(tfd, hpos+12, len, pos, clen);
			return;

		}

		hpos += 12;
		hpos += ((strlen(hpos) / 4) + 1) * 4;

	}

	terminate("resource %s is not found in %s", resource, themefile);

}

void main_l(const char *themefile) {

	char buf[32];
	unsigned char *header, *hpos;
	unsigned int *iheader;
	int headersize;
	unsigned long len, clen;
	FILE *tfd;

	tfd = fopen(themefile, "rb");
	if (tfd == NULL) terminate("Cannot open theme file %s", themefile);

	memset(buf, 0, 32);
	fread(buf, 1, 32, tfd);
	if (strncmp(buf, PBTSIGNATURE, strlen(PBTSIGNATURE)) != 0) terminate("%s is not a PocketBook theme file");
	headersize = *((int *) (buf+16));
	header = malloc(headersize);
	iheader = (int *) header;
	fseek(tfd, 0, SEEK_SET);
	fread(header, 1, headersize, tfd);

	hpos = header+32;

	printf("resource                 size     compressed\n");
	printf("--------------------------------------------\n");
	printf("<theme.cfg>          %8i      %8i\n", iheader[5], iheader[7]);

	while (hpos < header+headersize) {

		len = *((int *) hpos);
		clen = *((int *) (hpos+8));
		printf("%-20s %8i      %8i\n", hpos+12, (int)len, (int)clen);

		hpos += 12;
		hpos += ((strlen(hpos) / 4) + 1) * 4;

	}
	printf("\n");

}

int main(int argc, char **argv) {


	if (argc < 3){ usage(argv);
	}

	if (strcmp(argv[1], "-c") == 0) {
		if (argc < 4) usage(argv);
		main_c(argv[2], argc-3, &argv[3]);
	} else if (strcmp(argv[1], "-t") == 0) {
		if (argc < 4) usage(argv);
		main_t(argv[2], argv[3], argc-4, &argv[4]);
	} else if (strcmp(argv[1], "-u") == 0) {
		if (argc < 4) usage(argv);
		main_u(argv[2], argv[3]);
	} else if (strcmp(argv[1], "-l") == 0) {
		main_l(argv[2]);
	} else {
		usage(argv);
	}

	return 0;


}
